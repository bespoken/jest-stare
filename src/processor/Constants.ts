import * as chalk from "chalk";

/**
 * Contains Constants constants
 * @export
 * @class Constants
 */
export class Constants {

    /**
     * Name (same as package.json)
     * @static
     * @memberof Constants
     */
    public static readonly NAME = "jest-stare";

    /**
     * Logo for messaging
     * @static
     * @memberof Constants
     */
    public static readonly LOGO = chalk.default.green("**  ") + chalk.default.green("jest") + chalk.default.yellow("-") + chalk.default.red("stare");

    /**
     * Suffix for logo
     * @static
     * @memberof Constants
     */
    public static readonly SUFFIX = chalk.default.green("\t**");

    /**
     * Default jest-stare results directory
     * @static
     * @memberof Constants
     */
    public static readonly DEFAULT_RESULTS_DIR = "./" + Constants.NAME;

    /**
     * File extension for HTML files
     * @static
     * @memberof Constants
     */
    public static readonly HTML_EXTENSION = ".html";


    /**
     * Main HTML index file for report
     * @static
     * @memberof Constants
     */
    public static readonly MAIN_HTML = "index" + Constants.HTML_EXTENSION;

    /**
     * jest-stare render JS file
     * @static
     * @memberof Constants
     */
    public static readonly JEST_STARE_JS = "view.js";

    /**
     * Message for reporter run of jest-stare
     * @static
     * @memberof Constants
     */
    public static readonly REPORTER_WRITTING = " will write each completed run to ";

    /**
     * Raw JSON response results
     * @static
     * @memberof Constants
     */
    public static readonly RESULTS_RAW = "jest-results.json";

    /**
     * jest-stare custom css
     * @static
     * @memberof Constants
     */
    public static readonly JEST_STARE_CSS = Constants.NAME + ".css";

    /**
     * Template HTML file
     * @static
     * @memberof Constants
     */
    public static readonly TEMPLATE_HTML = "template.html";

    /**
     * Template HTML file
     * @static
     * @memberof Constants
     */
    public static readonly TEMPLATE_INLINE_SOURCE_HTML = "templateInlineSource.html";

    /**
     * Base dir to contain css files
     * @static
     * @memberof Constants
     */
    public static readonly CSS_DIR = "css/";

    /**
     * Base dir to contain js files
     * @static
     * @memberof Constants
     */
    public static readonly JS_DIR = "js/";

    /**
     * Base dir to contain image snapshot diff files
     * @static
     * @memberof Constants
     */
    public static readonly IMAGE_SNAPSHOT_DIFF_DIR = "image_snapshot_diff/";

    /**
     * jest option for test results processors
     * @static
     * @memberof Constants
     */
    public static readonly TEST_RESULTS_PROCESSOR = "--testResultsProcessor";

    /**
     * jest option for reporters
     * @static
     * @memberof Constants
     */
    public static readonly REPORTERS = "--reporters";

    /**
     * Log content after logo
     * @static
     * @memberof Constants
     */
    public static readonly LOG_MESSAGE = ": wrote output report to ";

    /**
     * Merging content message
     * @static
     * @memberof Constants
     */
    public static readonly MERGE_MESSAGE = ": will merge with ";

    /**
     * Error for whenever called without test results
     * @static
     * @memberof Constants
     */
    public static readonly NO_INPUT = Constants.NAME + " was called without input results";

    /**
     * Error for whenever the CLI is called without an input file is specified containing jest JSON
     * @static
     * @memberof Constants
     */
    public static readonly NO_CLI_INPUT = Constants.NAME + " CLI was called without input JSON file to read";

    /**
     * Notification called with programmatic config
     * @static
     * @memberof Constants
     */
    public static readonly OVERRIDE_JEST_STARE_CONFIG = Constants.NAME + " was called with programmatic config";
}
