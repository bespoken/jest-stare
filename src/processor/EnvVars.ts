import { IJestStareConfig } from "./doc/IJestStareConfig";
import { Logger } from "../utils/Logger";
import { EnvVarService } from "../utils/EnvVarService";

/**
 * Service for reading environmental variables
 * @export
 * @class EnvVars
 */
export class EnvVars {

    /**
     * Prefix for all jest-stare related environmental variables
     * @static
     * @type {string}
     * @memberof EnvVars
     */
    public static readonly ENV_PREFIX: string = "JEST_STARE_";

    /**
     * Creates an instance of EnvVars.
     * @memberof EnvVars
     */
    constructor(private mEnvSrv = new EnvVarService(EnvVars.ENV_PREFIX)) { }

    /**
     * Read all jest-stare related environmental variables
     * @returns {IJestStareConfig} - jest stare config from env vars
     * @memberof EnvVars
     */
    public read(): IJestStareConfig {
        const additionalResultsProcessorsValue = this.mEnvSrv.readEnvValue("ADDITIONAL_RESULTS_PROCESSORS");
        let additionalResultsProcessors: string[];
        if (additionalResultsProcessorsValue != null) {
            try {
                additionalResultsProcessors = JSON.parse(additionalResultsProcessorsValue);
            } catch (e) {
                Logger.get.error("Could not parse additional results processors value." +
                    "It should be a JSON string of an array of strings, like the following: " +
                    JSON.stringify(["jest-html-reporter"]) + "\n You specified: " + additionalResultsProcessorsValue);
            }
        }
        return {
            resultDir: this.mEnvSrv.readEnvValue("RESULT_DIR"),
            resultJson: this.mEnvSrv.readEnvValue("RESULT_JSON"),
            resultHtml: this.mEnvSrv.readEnvValue("RESULT_HTML"),
            log: this.mEnvSrv.readBoolEnvValue("LOG"),
            jestStareConfigJson: this.mEnvSrv.readEnvValue("CONFIG_JSON"),
            jestGlobalConfigJson: this.mEnvSrv.readEnvValue("GLOBAL_CONFIG_JSON"),
            coverageLink: this.mEnvSrv.readEnvValue("COVERAGE_LINK"),
            report: this.mEnvSrv.readBoolEnvValue("REPORT"),
            reportTitle: this.mEnvSrv.readEnvValue("REPORT_TITLE"),
            reportHeadline: this.mEnvSrv.readEnvValue("REPORT_HEADLINE"),
            reportSummary: this.mEnvSrv.readBoolEnvValue("REPORT_SUMMARY"),
            additionalResultsProcessors,
            disableCharts: this.mEnvSrv.readBoolEnvValue("DISABLE_CHARTS"),
            hidePassing: this.mEnvSrv.readBoolEnvValue("HIDE_PASSING"),
            hideFailing: this.mEnvSrv.readBoolEnvValue("HIDE_FAILING"),
            hidePending: this.mEnvSrv.readBoolEnvValue("HIDE_PENDING")
        };
    }

    /**
     * Combine jest stare config from package json and environmental variables,
     * with environmental variables taking precedence
     * @param packageJsonConfig
     * @param {IJestStareConfig} packageJsonConfig - package json config
     * @param {IJestStareConfig} envConfig - env based config
     * @returns {IJestStareConfig} - merged config
     * @memberof EnvVars
     */
    public resolve(packageJsonConfig: IJestStareConfig, envConfig: IJestStareConfig): IJestStareConfig {
        const mergedConfig: IJestStareConfig = {};

        if (envConfig.resultDir != null || packageJsonConfig.resultDir != null) {
            mergedConfig.resultDir = envConfig.resultDir == null ? packageJsonConfig.resultDir : envConfig.resultDir;
        }

        if (envConfig.resultJson != null || packageJsonConfig.resultJson != null) {
            mergedConfig.resultJson = envConfig.resultJson == null ? packageJsonConfig.resultJson : envConfig.resultJson;
        }

        if (envConfig.resultHtml != null || packageJsonConfig.resultHtml != null) {
            mergedConfig.resultHtml = envConfig.resultHtml == null ? packageJsonConfig.resultHtml : envConfig.resultHtml;
        }

        if (envConfig.log != null || packageJsonConfig.log != null) {
            mergedConfig.log = envConfig.log == null ? packageJsonConfig.log : envConfig.log;
        }

        if (envConfig.report != null || packageJsonConfig.report != null) {
            mergedConfig.report = envConfig.report == null ? packageJsonConfig.report : envConfig.report;
        }

        if (envConfig.reportTitle != null || packageJsonConfig.reportTitle != null) {
            mergedConfig.reportTitle = envConfig.reportTitle == null ? packageJsonConfig.reportTitle : envConfig.reportTitle;
        }

        if (envConfig.reportHeadline != null || packageJsonConfig.reportHeadline != null) {
            mergedConfig.reportHeadline = envConfig.reportHeadline == null ? packageJsonConfig.reportHeadline : envConfig.reportHeadline;
        }

        if (envConfig.reportSummary != null || packageJsonConfig.reportSummary != null) {
            mergedConfig.reportSummary = envConfig.reportSummary == null ? packageJsonConfig.reportSummary : envConfig.reportSummary;
        }

        if (envConfig.jestStareConfigJson != null || packageJsonConfig.jestStareConfigJson != null) {
            mergedConfig.jestStareConfigJson =
                envConfig.jestStareConfigJson == null ? packageJsonConfig.jestStareConfigJson : envConfig.jestStareConfigJson;
        }

        if (envConfig.jestGlobalConfigJson != null || packageJsonConfig.jestGlobalConfigJson != null) {
            mergedConfig.jestGlobalConfigJson =
                envConfig.jestGlobalConfigJson == null ? packageJsonConfig.jestGlobalConfigJson : envConfig.jestGlobalConfigJson;
        }

        if (envConfig.coverageLink != null || packageJsonConfig.coverageLink != null) {
            mergedConfig.coverageLink = envConfig.coverageLink == null ? packageJsonConfig.coverageLink : envConfig.coverageLink;
        }

        if (envConfig.additionalResultsProcessors != null || packageJsonConfig.additionalResultsProcessors != null) {
            mergedConfig.additionalResultsProcessors =
                envConfig.additionalResultsProcessors == null ? packageJsonConfig.additionalResultsProcessors : envConfig.additionalResultsProcessors;
        }

        if (envConfig.disableCharts != null || packageJsonConfig.disableCharts != null) {
            mergedConfig.disableCharts =
                envConfig.disableCharts == null ? packageJsonConfig.disableCharts : envConfig.disableCharts;
        }

        if (envConfig.hidePassing != null || packageJsonConfig.hidePassing != null) {
            mergedConfig.hidePassing =
                envConfig.hidePassing == null ? packageJsonConfig.hidePassing : envConfig.hidePassing;
        }

        if (envConfig.hideFailing != null || packageJsonConfig.hideFailing != null) {
            mergedConfig.hideFailing =
                envConfig.hideFailing == null ? packageJsonConfig.hideFailing : envConfig.hideFailing;
        }


        if (envConfig.hidePending != null || packageJsonConfig.hidePending != null) {
            mergedConfig.hidePending =
                envConfig.hidePending == null ? packageJsonConfig.hidePending : envConfig.hidePending;
        }
        return mergedConfig;
    }
}
