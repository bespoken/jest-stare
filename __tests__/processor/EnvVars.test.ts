import { EnvVars } from "../../src/processor/EnvVars";

describe("EnvVars tests", () => {
    it("should have an empty config for no env variables set", () => {
        const response = new EnvVars().read();
        expect(response).toMatchSnapshot();
    });

    it("should have a complete config for all env variables set", () => {
        const JEST_STARE_RESULT_DIR = "JEST_STARE_RESULT_DIR";
        const JEST_STARE_RESULT_JSON = "JEST_STARE_RESULT_JSON";
        const JEST_STARE_RESULT_HTML = "JEST_STARE_RESULT_HTML";
        const JEST_STARE_REPORT_TITLE = "JEST_STARE_REPORT_TITLE";
        const JEST_STARE_REPORT_HEADLINE = "JEST_STARE_REPORT_HEADLINE";
        const JEST_STARE_REPORT_SUMMARY = "JEST_STARE_REPORT_SUMMARY";
        const JEST_STARE_LOG = "JEST_STARE_LOG";
        const JEST_STARE_MERGE = "JEST_STARE_MERGE";
        const JEST_STARE_CONFIG_JSON = "JEST_STARE_CONFIG_JSON";
        const JEST_STARE_COVERAGE_LINK = "JEST_STARE_COVERAGE_LINK";
        const JEST_STARE_ADDITIONAL_RESULTS_PROCESSORS = "JEST_STARE_ADDITIONAL_RESULTS_PROCESSORS";
        const JEST_STARE_DISABLE_CHARTS = "JEST_STARE_DISABLE_CHARTS";
        const JEST_STARE_HIDE_PASSING = "JEST_STARE_HIDE_PASSING";
        const JEST_STARE_HIDE_FAILING = "JEST_STARE_HIDE_FAILING";
        const JEST_STARE_INLINE_SOURCE = "JEST_STARE_INLINE_SOURCE";
        const JEST_STARE_GENERATE_PDF = "JEST_STARE_GENERATE_PDF";
        process.env[JEST_STARE_RESULT_DIR] = "some/dir";
        process.env[JEST_STARE_RESULT_JSON] = "some.json";
        process.env[JEST_STARE_RESULT_HTML] = "some.html";
        process.env[JEST_STARE_REPORT_TITLE] = "my title!";
        process.env[JEST_STARE_REPORT_HEADLINE] = "my title";
        process.env[JEST_STARE_REPORT_SUMMARY] = "false";
        process.env[JEST_STARE_LOG] = "false";
        process.env[JEST_STARE_MERGE] = "1";
        process.env[JEST_STARE_CONFIG_JSON] = "true";
        process.env[JEST_STARE_COVERAGE_LINK] = "www.walmart.com";
        process.env[JEST_STARE_ADDITIONAL_RESULTS_PROCESSORS] = "[\"one\", \"two\", \"three\"]";
        process.env[JEST_STARE_DISABLE_CHARTS] = "false";
        process.env[JEST_STARE_HIDE_PASSING] = "true";
        process.env[JEST_STARE_HIDE_FAILING] = "true";
        process.env[JEST_STARE_INLINE_SOURCE] = "false";
        process.env[JEST_STARE_GENERATE_PDF] = "false";
        
        const response = new EnvVars().read();
        expect(response).toMatchSnapshot();
        const newResponse = new EnvVars().resolve({}, response);
        expect(newResponse).toMatchSnapshot();
    });
});
