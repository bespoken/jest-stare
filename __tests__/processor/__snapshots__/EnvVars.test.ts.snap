// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvVars tests should have a complete config for all env variables set 1`] = `
Object {
  "additionalResultsProcessors": Array [
    "one",
    "two",
    "three",
  ],
  "coverageLink": "www.walmart.com",
  "disableCharts": false,
  "generatePdf": false,
  "generatePdfToken": undefined,
  "hideFailing": true,
  "hidePassing": true,
  "hidePending": undefined,
  "inlineSource": false,
  "jestGlobalConfigJson": undefined,
  "jestStareConfigJson": "true",
  "log": false,
  "report": undefined,
  "reportHeadline": "my title",
  "reportSummary": false,
  "reportTitle": "my title!",
  "resultDir": "some/dir",
  "resultHtml": "some.html",
  "resultJson": "some.json",
}
`;

exports[`EnvVars tests should have a complete config for all env variables set 2`] = `
Object {
  "additionalResultsProcessors": Array [
    "one",
    "two",
    "three",
  ],
  "coverageLink": "www.walmart.com",
  "disableCharts": false,
  "generatePdf": false,
  "hideFailing": true,
  "hidePassing": true,
  "inlineSource": false,
  "jestStareConfigJson": "true",
  "log": false,
  "reportHeadline": "my title",
  "reportSummary": false,
  "reportTitle": "my title!",
  "resultDir": "some/dir",
  "resultHtml": "some.html",
  "resultJson": "some.json",
}
`;

exports[`EnvVars tests should have an empty config for no env variables set 1`] = `
Object {
  "additionalResultsProcessors": undefined,
  "coverageLink": undefined,
  "disableCharts": undefined,
  "generatePdf": undefined,
  "generatePdfToken": undefined,
  "hideFailing": undefined,
  "hidePassing": undefined,
  "hidePending": undefined,
  "inlineSource": undefined,
  "jestGlobalConfigJson": undefined,
  "jestStareConfigJson": undefined,
  "log": undefined,
  "report": undefined,
  "reportHeadline": undefined,
  "reportSummary": undefined,
  "reportTitle": undefined,
  "resultDir": undefined,
  "resultHtml": undefined,
  "resultJson": undefined,
}
`;
